pkg/mutation/system.go

```diff
@@ -1,20 +1,173 @@
 type System struct {
-	// schemaDB schema.DB commented to please the linter, will uncomment in the implementation
-	mutators []Mutator
+	schemaDB        SchemaDB
+	orderedMutators []Mutator
+	mutatorsMap     map[ID]Mutator
 	sync.RWMutex
 }
 
 // NewSystem initializes an empty mutation system
 func NewSystem() *System {
 	return &System{
-		mutators: make([]Mutator, 0),
+		orderedMutators: make([]Mutator, 0),
+		mutatorsMap:     make(map[ID]Mutator),
+	}
+}
+
+// Upsert updates or insert the given object, and returns
+// an error in case of conflicts
+func (s *System) Upsert(m Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	current, ok := s.mutatorsMap[m.ID()]
+	if ok && !m.HasDiff(current) {
+		return nil
+	}
+
+	toAdd := m.DeepCopy()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := toAdd.(MutatorWithSchema); ok {
+		err := s.schemaDB.Upsert(withSchema)
+		if err != nil {
+			return errors.Wrapf(err, "Schema upsert failed")
+		}
+	}
+
+	s.mutatorsMap[toAdd.ID()] = toAdd
+
+	i := sort.Search(len(s.orderedMutators), func(i int) bool {
+		return greaterOrEqual(s.orderedMutators[i], toAdd)
+	})
+
+	if i == len(s.orderedMutators) { // Adding to the bottom of the list
+		s.orderedMutators = append(s.orderedMutators, toAdd)
+		return nil
+	}
+
+	found := equal(s.orderedMutators[i], toAdd)
+	if found {
+		s.orderedMutators[i] = toAdd
+		return nil
+	}
+
+	s.orderedMutators = append(s.orderedMutators, nil)
+	copy(s.orderedMutators[i+1:], s.orderedMutators[i:])
+	s.orderedMutators[i] = toAdd
+	return nil
+}
+
+// Mutate applies the mutation in place to the given object
+func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) error {
+	s.RLock()
+	defer s.RUnlock()
+
+	maxIterations := len(s.orderedMutators) + 1
+
+	old := obj.DeepCopy()
+
+	for i := 0; i < maxIterations; i++ {
+		for _, m := range s.orderedMutators {
+			if m.Matches(obj, ns) {
+				err := m.Mutate(obj)
+				if err != nil {
+					return errors.Wrapf(err, "Mutation failed for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+				}
+			}
+		}
+		if cmp.Equal(old, obj) {
+			return nil
+		}
+		old = obj.DeepCopy()
+	}
+
+	return fmt.Errorf("Mutation not converging for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+}
+
+// Remove removes the mutator from the mutation system
+func (s *System) Remove(m Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if _, ok := s.mutatorsMap[m.ID()]; !ok {
+		return nil
+	}
+
+	err := s.schemaDB.Remove(m.ID())
+	if err != nil {
+		return errors.Wrapf(err, "Schema remove failed")
+	}
+
+	i := sort.Search(len(s.orderedMutators), func(i int) bool {
+		res := equal(s.orderedMutators[i], m)
+		if res {
+			return true
+		}
+		return greaterOrEqual(s.orderedMutators[i], m)
+	})
+
+	found := equal(s.orderedMutators[i], m)
+
+	// The map is expected to be in sync with the list, so if we don't find it
+	// we return an error.
+	if !found {
```

Comment:
How should we recover from this error?

Reply:
If this error happens, the current state is the desired one (both the map and the slice do not contain the given element), so in theory there is nothing to recover from.
On the other hand, this should be raised in some way, because it means that there's a bug and at some point the map and the slice were not aligned. I'd prefer this failing (even brutally) than silently getting recovered.

Reply:
Fair enough, though we will be hard down unless we delete the mutator from the map before returning this error:

`Remove()` will return an error, mutators map will retain the reference, the reconciler will attempt to delete the mutator again and we enter an infinite loop.

Definitely okay with returning an error, but we should figure out how to make sure the next iteration succeeds.

Reply:
Sorry I mis-interpreted the comment (or, my code :-) )
I'll remove the item from the map anyway, so at least the state is consistent and in the next iteration it will pass.

Reply:
ack, looks good now!